app/controllers/admin/api_docs/base_controller.rb

```diff
@@ -0,0 +1,122 @@
+class Admin::ApiDocs::BaseController < FrontendController
+  before_action :find_api_docs, only: %i[destroy edit update show preview toggle_visible]
+  before_action :find_service_apis_collection, only: %i[new edit]
+  before_action :deny_on_premises_for_master
+
+  def preview
+    if api_docs_service.specification.swagger?
+      respond_to do |format|
+        format.html { render 'swagger' }
+        format.json { render json: swagger_spec }
+      end
+    else
+      @host = current_account.domain
+      @resource = api_docs_service.system_name
+      @spec = ApiDocs::Service.spec_for api_docs_service
+      render 'active_docs'
+    end
+  end
+
+  def toggle_visible
+    api_docs_service.toggle! :published
+
+    message = api_docs_service.published? ? 'published' : 'unpublished'
+    redirect_to preview_admin_api_docs_service_path(api_docs_service), notice: "Spec #{api_docs_service.name} #{message}"
+  end
+
+  def show
+    specification = api_docs_service.specification
+    if specification.swagger_2_0?
+      # TODO: this is temporary until we get swagger-ui to load swagger-2.0
+      response.headers['X-Frame-Options'] = 'SAMEORIGIN'
+      json = JSON.pretty_generate specification.as_json
+    else
+      translator = ThreeScale::Swagger::Translator.translate! @api_docs_service.body
+      json = translator.as_json
+    end
+
+    respond_to do | format |
+      format.json { render json: json }
+    end
+  end
+
+  def new
+    @api_docs_service = current_scope.api_docs_services.new
+  end
+
+  def edit; end
+
+  def update
+    respond_to do |format|
+      if api_docs_service.update(api_docs_params, without_protection: true)
+        message = 'ActiveDocs Spec was successfully updated.'
+        format.html { redirect_to(preview_admin_api_docs_service_path(api_docs_service), notice: message) }
+        format.js { render js: "jQuery.flash.notice('#{message}')" }
+      else
+        flash[:error] = api_docs_service.errors.full_messages.join(' ')
+        format.html { render :edit }
+        format.js {}
+      end
+    end
+  end
+
+  def index
+    @api_docs_services = current_scope.api_docs_services.page(params[:page])
+  end
+
+  def create
+    @api_docs_service = current_scope.api_docs_services.new(api_docs_params(:system_name), without_protection: true)
+    if @api_docs_service.save
+      redirect_to(preview_admin_api_docs_service_path(@api_docs_service), notice: 'ActiveDocs Spec was successfully saved.')
+    else
+      flash[:error] = @api_docs_service.errors.full_messages.join(' ')
+      render :new
+    end
+  end
+
+  def destroy
+    api_docs_service.destroy
+    redirect_to admin_api_docs_services_path, notice: 'ActiveDocs Spec was successfully deleted.'
+  end
+
+  private
+
+  attr_reader :api_docs_service
+
+  def find_service_apis_collection
+    @service_apis = current_account.services.accessible
+  end
+
+  def current_scope
+    raise NotImplementedError, "#{__method__} not implemented in #{self.class}"
```

Comment:
See https://github.com/3scale/porta/pull/72/files#r224312319

Reply:
CodeClimate didn't complain about this one but about the other one inheriting from NotImplementedError. So I will change the other one 👍 but what do you mean in this one?

Reply:
Well `NotImplementedError` is not really what should be used.
It is meant for OS level implementation

Reply:
Ok so you sent me http://chrisstump.online/2016/03/23/stop-abusing-notimplementederror/ in Slack which seems to say that it is better not writing any method of this sort at all and it will automatically raise NoMethodError...

However, Michal senior told me this once:
![image](https://user-images.githubusercontent.com/11318903/46787939-3e7cdb00-cd39-11e8-9d3e-96584b2d8569.png)
And Sandi Metz said the same in her book:
https://gist.github.com/speric/31ae0987d21eac1d4f87 (do CTRL+F and NotImplementedError)

Reply:
Actually there is a comment in the blog that you sent me:

> Sandi Metz seems to suggest that raising some type of error – anything at all – is better than the no method error situation which arises if the super class relies on the sub-class to implementing the method exclusively. In other words, better to be explicitly clear – so that things are obvious at a glance.
> 
> she suggests that things might be obvious for the immediate developer, but a future one who doesn’t undersetand the code base, may be confused if such an exception is not raised. i share this view, even though tehcnically, the exception may not be the correct one.

Reply:
![image](https://user-images.githubusercontent.com/11318903/46788460-6c165400-cd3a-11e8-9f0c-c94f040114cc.png)
![image](https://user-images.githubusercontent.com/11318903/46788480-73d5f880-cd3a-11e8-8a84-54697434c0ae.png)

Reply:
Sure but I disagree with her using the class NotImplementedError. Which should be used as documentation says: 

> Raised when a feature is not implemented on the current platform. For example, methods depending on the fsync or fork system calls may raise this exception if the underlying operating system or Ruby runtime does not support them.
>
> Note that if fork raises a NotImplementedError, then respond_to?(:fork) returns false.

Better use a subclass of NoMethodError (or bare NoMethodError)